### PR TITLE
chore: add carousel spec

### PR DESCRIPTION
**Pull request by SethDonohue**
# Description

Adds web component spec for Carousel component

## Issue type checklist

- [x] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [ ] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

## Process & policy checklist

- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: specs/carousel/carousel.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            slot="play-toggle"
+            @click="${x => x.togglePlay()}"
+            class="previous-flipper"
+        >
+            <svg class="play-svg">
+            <path d="example path..."></path>
+            </svg>
+        </button>
+
+        <div role="tablist" class="tablist" aria-label="A carousel of items">
```
Comment (eljefe223) [inline specs/carousel/carousel.md]: I just started looking into a web component version of Tabs might be good to sync up here.

File: specs/carousel/carousel.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+*Slots:*
+- `default` - slot for slides content // TODO: IS this the right way?
+- `play-toggle` - the control to toggle if the slides rotate or not
+- `next-flipper` - the control used to rotate to the next slide
+- `previous-flipper` - the control used to rotate to the previous slide
+- `tablist` - the control to select a certain slide
+
+*Events*
+- `on-pause` - callback fired when the rotation is stopped, such as when keyboard focus or hover is on any element of the carousel, or the *play-toggle* is set to off.
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: we typically don't use the "on-" convention for event names; I think the first would just be "pause". The [ARIA carousel spec](https://www.w3.org/TR/wai-aria-practices/#carousel) uses the term "rotate" a lot - thoughts on using "rotate" for the event name instead of "active-id-update"?

Comment (nicholasrice) [reply specs/carousel/carousel.md]: We might also want a "play" event also? Or maybe this is all surfaced under a general "autorotate" event?

File: specs/carousel/carousel.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                class="carousel-item ${x => x.getActiveIndex() === index ? 'active' : ''}"
+                role="group"
+                aria-roledescription="slide"
+                aria-label="${x => x.getActiveIndex()} of ${x => x.getItemCount()}"
+            >
+                ${item}
+            </div>`
+            }).join('')
+        }
+        <!-- Solution 2 -->
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: Only having one item in the DOM might make slide transitions difficult or impossible to achieve.

File: specs/carousel/carousel.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        </div>
+    </div>
+    <slot name="previous-flipper"> SEE DEFAULT </slot>
+    <slot name="tab-list">SEE DEFAULT </slot>
+    </div>
+</template>
+```
+
+**Default Flipper**
+```html
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: Having default buttons poses an interesting question. These elements will need localized text labels to communicate to AT what they are for. Typically we don't bake localized strings into elements though.

File: specs/carousel/carousel.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    <div tabindex="-1" role="tab" class="tablist-tab"></div>
+    <div tabindex="-1" role="tab" class="tablist-tab"></div>
+</div>
+```
+
+**Implementation**
+```html
+<body>
+    <fast-carousel
+        // TODO: CHECK is this a way to apply the items?
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: One option for providing the items would be to use slots. `` The control can then determine what to do with all of the slotted content.

File: specs/carousel/carousel.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        item-0='<img src="...image1.src">'
+        item-1='<img src="...image2.src">'
+        item-2='<img src="...image3.src">'
+    >
+        <button
+            slot="next-flipper"
+            @click="${x => x.next()}" // TODO: How do I attach the carousel.next() function to a slotted item?
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: You could use event delegation for this. The host element would listen for click events and then dispatch an action based on which element was clicked. For example, if the composed path comes from an element that has the `[slot="next-flipper"]` attribute then we do work to render the next slide.

File: specs/carousel/carousel.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+---
+
+## Implementation
+
+### States
+
+*Key component states, valid state transitions, and how interactions trigger a state transition.*
+
+- `isPaused` - if the carousel slide rotation is paused
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: you could probably shorten this to just `paused` because it is a boolean.

File: specs/carousel/carousel.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+- `activeId` - the id of the active slide
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: It might be useful to just store a reference to the active slide element itself.

File: specs/carousel/carousel.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+The carousel should align to the design pattern and interaction model provided by the W3C: https://w3c.github.io/aria-practices/#carousel
+
+### Globalization
+
+The component may have to change the order of rotation for slides as an RTL users sequence expectations is reversed from LTR. This would require that the flippers and tablist be functionally reversed. The tablist would also have to be visually reversed with the first tab item starting on the right of the tablist.
+
+### Test Plan
+
+
```
Comment (nicholasrice) [inline specs/carousel/carousel.md]: See note about using slots to provide the items regarding this perf comment.

File: specs/carousel/carousel.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+*Component name:*
+- `fast-carousel`
+
+*Attributes:*
+- `autoplay` - boolean value to toggle if the carousel plays on load or not (defaults to false)
+- `autoplay-interval` - a number representing the autoplay interval in milliseconds
+- `loop` - boolean value to toggle if the carousel loops slides at the beginning or end (defaults to true)
+- `paused` - whether the rotation is paused or not
+- `aria-roledescription` - set to *carousel* for overall component, and set to *slide* for each slide
```
Comment (eljefe223) [inline specs/carousel/carousel.md]: This shounds like it does not need to be an attribute if it always needs to be set to carousel or slide

File: specs/carousel/carousel.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+>
+    <svg class="play-svg">
+        <path d="example path..."></path>
+    </svg>
+</button>
+```
+
+**Default Tab List**
+```html
+<div role="tablist" class="tablist" aria-label="A carousel of items">
```
Comment (eljefe223) [inline specs/carousel/carousel.md]: We will want to use whatever we decide for `fast-tabs` here. I just started looking into that spec. Maybe we should sync up on this.

Comment (chrisdholt) [reply specs/carousel/carousel.md]: I think a sync might be helpful here, but I'm not 100% sure we need to map to `tabs`. There is a working draft for carousel in the W3C as a control: https://w3c.github.io/aria-practices/#carousel

I do have some thoughts on the above spec though as it seems to include several interaction models that conflict and could make implementing difficult or confusing for end users.

File: specs/carousel/carousel.md (hunk 2 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Yeah I need to put much more thought into this part. I like your ideas and will add them in.

File: specs/carousel/carousel.md (hunk 3 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Yes, a typical carousel has a few so I'll look at updating this.

File: specs/carousel/carousel.md (hunk 4 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Good thoughts. I assume we want implementors/devs to be able to use their own flippers, so I'll add this concern to the list and see if I can come up with something.

Comment (chrisdholt) [reply specs/carousel/carousel.md]: I think one way to approach this is how we default the control. For instance, the default carousel would like quite a bit like our current implementation with sequence indicators and flippers which are hidden from AT. This way, the default flippers are hidden from AT by design. To unhide them you would need to pass your own flippers. We could also easily ensure that buttons match the styling by using the slotted pseudo selector to assign the styles to any button passed - `::slotted(button)`. I think there's a lot of room to creatively solve this.

File: specs/carousel/carousel.md (hunk 11 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Yes, I saw that the old one uses tabs, but feel like it may not be necessary. I'll address some of these other issues and then sync with you both on the above.

File: specs/carousel/carousel.md (hunk 3 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Updated Solution 2 to have 3 slides

File: specs/carousel/carousel.md (hunk 5 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Like using a querySelector([slot="play-toggle"]) to grab all the items. Sounds good!

File: specs/carousel/carousel.md (hunk 6 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Updated to have the click handler on the host element.

File: specs/carousel/carousel.md (hunk 8 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Added a reference prop

File: specs/carousel/carousel.md (hunk 4 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md:143]: Do we want buttons to have a certain styling, or do we want the author/implementor to pass all that in?
Maybe a sync on this idea and others is a good idea. I'll get that going soon

File: specs/carousel/carousel.md (hunk 5 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: Updated to use "item" slot and then the controller can handle all the items given.

File: specs/carousel/carousel.md [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+** See the [W3C Specification](https://w3c.github.io/aria-practices/#wai-aria-roles-states-and-properties-4) for requirements and details.
+
+*CSS custom properties:*
+- ???
+
+*Slots:*
+- `default` - slot for slides content // TODO: IS this the right way?
+- `play-toggle` - the control to toggle if the slides rotate or not
+- `next-flipper` - the control used to rotate to the next slide
```
Comment (nicholasrice) [inline specs/carousel/carousel.md:60]: I kind of think "flipper" isn't the right term here. That is the name of the control *we* use for this, but it's really just a button. Thoughts on changing this to somthing like `next-button` / `previous-button`?

File: specs/carousel/carousel.md (hunk 4 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md:143]: @chrisdholt Does your fast-flipper address all of this? I took the defaults out for the spec and actually have the **Implementation** using fast-flipper.

Comment (chrisdholt) [reply specs/carousel/carousel.md]: There's likely a couple ways to handle it:

1) Take a dependency on Flipper.

2) Support default buttons in the slot which use the fast-flipper styles. Doing this would allow reuse if someone wanted to pass the fast-flipper themselves and handle telemetry, etc.

3) Some other way?

Comment (SethDonohue) [reply specs/carousel/carousel.md]: > @chrisdholt Does your fast-flipper address all of this? I took the defaults out for the spec and actually have the **Implementation** using fast-flipper.

Correction: I added fast-flipper as the default for the next-button and previous-button slots in the template.

File: specs/carousel/carousel.md (hunk 11 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md]: I have added a simple tablist example in the **implementation** of the spec.

File: specs/carousel/carousel.md [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            <div tabindex="0" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+        </div>
+    >
+    </fast-carousel>
+</body>
+```
```
Comment (EisenbergEffect) [inline specs/carousel/carousel.md:177]: It doesn't look like we're using the default slot. Can we put the items in the default slot?

Also, is it possible to generate the tab-list if one isn't provided? This would need to be kept in sync with the actual slotted items, but that's not too hard (I've got an idea for a library feature to help with this). These could be generated as fallback content for the tab-list slot.

Comment (SethDonohue) [reply specs/carousel/carousel.md:177]: Changed to use the default.

Yeah I think we are still discussing how tab-list will be used per the W3 Carousel spec as there are some different interaction models that could be used.
I still assume it'd be best to have a default. What's your idea for the library feature to help?

Comment (EisenbergEffect) [reply specs/carousel/carousel.md:177]: I'm thinking of having a `@slotted` decorator that would give you the list of html elements that exist in a particular slot, keeping it up to date automatically as slotted content changes:

```TypeScript
export class FastCarousel extends FastElement {
  @slotted items: HTMLElement[]; // all the html elements in the default slot 
}
````

With this in place you could repeater over `items` and generate the tab list into the default slot. Any time the slotted items change, the repeat would automatically update the tab list for you.

The `@slotted` decorator would take an optional slot name and an optional element filter for more advances scenarios. It would default to the default slot and a filter that selected all nodes.

Comment (EisenbergEffect) [reply specs/carousel/carousel.md:177]: Another option is to do something like ref where you put a directive on your slot and it gives you a property with its content:

```
const template = html`
  <slot ${slotted('items')}></slot>
`;

File: specs/carousel/carousel.md [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+                </div>
+                <div
+                    class="carousel-item-next"
+                    role="group"
+                    aria-roledescription="slide"
+                    aria-label="${x.getNextIndex} of ${x.getItemCount}"
+                >
+                    ${x.getNextItem()}
+                </div>`
+        }
```
[the remark this answers is not in the document]
Comment (EisenbergEffect) [reply specs/carousel/carousel.md:113]: I don't think the code you have above is going to work. I'm not quite following why you have the nested lambda expression there.

I'd probably go with rendering all the items and using the classes to activate them. We do have a new directive called `compose` that's planned to handle more advanced dynamic and conditional rendering of different templates.

File: specs/carousel/carousel.md (hunk 13 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md:177]: To be clear the items are the slide content, and in this template the tab-list is for the slide picker controls. Maybe I should name that something different that tab-list?

Comment (SethDonohue) [reply specs/carousel/carousel.md:177]: > I'm thinking of having a `@slotted` decorator that would give you the list of html elements that exist in a particular slot, keeping it up to date automatically as slotted content changes:
> 
> ```ts
> export class FastCarousel extends FastElement {
>   @slotted items: HTMLElement[]; // all the html elements in the default slot 
> }
> ```
> 
> With this in place you could repeater over `items` and generate the tab list into the fallback content for tab list. Any time the slotted items change, the repeat would automatically update the tab list for you.
> 
> The `@slotted` decorator would take an optional slot name and an optional element filter for more advances scenarios. It would default to the default slot and a filter that selected all nodes.

I think this would be handy for what I am trying to do. I am currently having difficulty getting the nodes/html elements that are in the default slot, and then using them to render the slides, but I believe I am missing a key concept.

Comment (EisenbergEffect) [reply specs/carousel/carousel.md:177]: You might be accessing assignedNodes to early and need to also register for the slotchanged event. What I'm proposing would do all that work for you so you just have a nice array of nodes in the end, which you could repeat over to generate other UI. I can probably get to to work on this Monday or Tuesday of next week if this sounds good.

File: specs/carousel/carousel.md [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+- `autoplay-interval` - a number representing the autoplay interval in milliseconds
+- `loop` - boolean value to toggle if the carousel loops slides at the beginning or end (defaults to true)
+- `aria-labelledby` - optional based on implementation**
+- `aria-label` - optional, based on implementation**
+- `paused` - whether the rotation is paused or not
+
+** See the [W3C Specification](https://w3c.github.io/aria-practices/#wai-aria-roles-states-and-properties-4) for requirements and details.
+
+*CSS custom properties:*
+- ???
```
[the remark this answers is not in the document]
Comment (nicholasrice) [reply specs/carousel/carousel.md:51]: If we don't need any then we can just omit this section

File: specs/carousel/carousel.md (hunk 13 above)
Comment (SethDonohue) [reply specs/carousel/carousel.md:177]: Sorry on the delay, this definitely sounds good.

File: specs/carousel/carousel.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+---
+
+## Design
+
+### API
+
+*Component name:*
+- `fast-carousel`
+
+*Attributes & Properties:*
```
Comment (scomea) [inline specs/carousel/carousel.md:40]: Should we expose the current slide index?  And somehow enable authors to navigate to a specific slide somehow?

File: specs/carousel/carousel.md [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+
+*Slots:*
+- `default` - use default slot for slide content
+- `play-toggle` - the control to toggle if the slides rotate or not
+- `next-button` - the control used to rotate to the next slide
+- `previous-button` - the control used to rotate to the previous slide
+- `tab-list` - the control to select a certain slide
+
+*Events*
+- `change` - callback fired when the slide changes. This updates the `activeslideid` and passes a reference to the active slide. The play/pause state and `activeslideid` can then be read by the author from the reference element, e.g. event.target.paused, event.target.activeslideid
```
Comment (eljefe223) [inline specs/carousel/carousel.md:64]: In tabs we pass an reference to the `activetab` and then allow the app author to set the `activetabid`

File: specs/carousel/carousel.md [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+Notes:
+- Since the host element has the handleClick(), it would use event delegation to check what element was clicked on. Example: If the target was an element with a `[slot="next-button"]` it would trigger the controller method for nextSlide(). This way each element passed in via slot would not needs it's own onClick.
+
+```html
+<!-- shadow dom -->
+<div
+    class="carousel"
+    aria-roledescription="carousel"
+    tabindex="-1"
+    @click="${x => x.handleClick()}"
```
Comment (eljefe223) [inline specs/carousel/carousel.md:78]: If this is the shadow dom we wouldn't see the click handler here.

Comment (EisenbergEffect) [reply specs/carousel/carousel.md:78]: I think this will be solved by attaching the click handler to the host itself. That should let it see the clicks from the light dom child nodes. Good catch @eljefe223 !

File: specs/carousel/carousel.md (hunk 13 above)
Comment (EisenbergEffect) [reply specs/carousel/carousel.md:154]: FYI, the new `slotted` directive is in. Right now you'll have to filter the nodes to ensure you only get element nodes, which is a pain. I'll see if I can add the filter feature in soon.
